File: 621-task-scheduler/task-scheduler.java
class Pair{
    int remaining_count;
    int time;

    public Pair(int count, int time){
        this.remaining_count=count;
        this.time=time;
    }
}
    
class Solution {
    public int leastInterval(char[] tasks, int n) {
        
        int[] frequencyMap = new int[26];

        for(char task: tasks){
            frequencyMap[task-'A']++;
        }

        PriorityQueue<Integer> pq= new PriorityQueue<>((a,b)-> b-a);

        for(int val: frequencyMap){
            if(val>0){
                pq.add(val);
            }
        }

        Queue<Pair> cooldown= new LinkedList<>();

        int time=0;
        while(!pq.isEmpty() || !cooldown.isEmpty()){
           

            if(!pq.isEmpty()){
                int count= pq.poll();
                count-=1;
                if(count>0){
                    cooldown.offer(new Pair(count,time+n));
                }
            }
            if(!cooldown.isEmpty() && cooldown.peek().time==time){
                pq.offer(cooldown.poll().remaining_count);
            }
            time+=1;
        }
        return time;

    }
}

File: 621-task-scheduler/task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        frequencyMap=[0]*26

        for task in tasks:
            frequencyMap[ord(task)-ord('A')]+=1
        

        heap=[ - count for count in frequencyMap if count>0 ]

        heapq.heapify(heap)

        cooldown= deque()

        time=0
        while heap or cooldown:
            time+=1
            if heap:
                count=heapq.heappop(heap);
                count+=1
                if count<0:
                    cooldown.append((count,time+n))
            if cooldown and cooldown[0][1]==time:
                remaining_count,_ = cooldown.popleft()
                heapq.heappush(heap, remaining_count)
        
        return time


